src/Workspaces/Remote/Core/RemoteHostAssetSerialization.cs

```diff
@@ -78,17 +78,17 @@ static void WriteAsset(ObjectWriter writer, ISerializerService serializer, Solut
             using var _ = FlowControlHelper.TrySuppressFlow();
             return ReadDataSuppressedFlowAsync(pipeReader, solutionChecksum, objectCount, serializerService, cancellationToken);
 
-            static async ValueTask<ImmutableArray<object>> ReadDataSuppressedFlowAsync(
+            static async ValueTask<ImmutableArray<T>> ReadDataSuppressedFlowAsync(
                 PipeReader pipeReader, Checksum solutionChecksum, int objectCount, ISerializerService serializerService, CancellationToken cancellationToken)
             {
                 using var stream = await pipeReader.AsPrebufferedStreamAsync(cancellationToken).ConfigureAwait(false);
-                return ReadData(stream, solutionChecksum, objectCount, serializerService, cancellationToken);
+                return ReadData<T>(stream, solutionChecksum, objectCount, serializerService, cancellationToken);
             }
         }
 
-        public static ImmutableArray<object> ReadData(Stream stream, Checksum solutionChecksum, int objectCount, ISerializerService serializerService, CancellationToken cancellationToken)
+        public static ImmutableArray<T> ReadData<T>(Stream stream, Checksum solutionChecksum, int objectCount, ISerializerService serializerService, CancellationToken cancellationToken)
         {
-            using var _ = ArrayBuilder<object>.GetInstance(objectCount, out var results);
+            using var _ = ArrayBuilder<T>.GetInstance(objectCount, out var results);
```

Comment:
>using var _ = ArrayBuilder<T>.GetInstance(objectCount, out var results); [](http://example.com/codeflow?start=12&length=72)

nit: might be nice to just new up the array directly and use ImmutableCollectionsMarshal.AsImmutableArray to get the ImmutableArray.

No need to hit the ArrayBuilder pool, especially since this now would hit lots of different pools because of the generic.

Reply:
also, these arrays can exceed the default threshhold, which can cause resizing and throwing away the old values in the pool.

Reply:
will do.

Reply:
done.